compiler/rustc_middle/src/ty/fast_reject.rs

```diff
@@ -182,3 +184,188 @@ impl<D: Copy + Debug + Eq> SimplifiedTypeGen<D> {
         }
     }
 }
+
+/// Given generic arguments from an obligation and an impl,
+/// could these two be unified after replacing parameters in the
+/// the impl with inference variables.
+pub fn generic_args_may_unify(
+    obligation_arg: ty::GenericArg<'_>,
+    impl_arg: ty::GenericArg<'_>,
+) -> bool {
+    match (obligation_arg.unpack(), impl_arg.unpack()) {
+        // We don't fast reject based on regions for now.
+        (GenericArgKind::Lifetime(_), GenericArgKind::Lifetime(_)) => true,
+        (GenericArgKind::Type(obl), GenericArgKind::Type(imp)) => types_may_unify(obl, imp),
+        (GenericArgKind::Const(obl), GenericArgKind::Const(imp)) => consts_may_unify(obl, imp),
+        _ => bug!("kind mismatch: {obligation_arg} {impl_arg}"),
+    }
+}
+
+/// Given a type from an obligation and a type from an impl, could those
+/// two types be unified after replacing the generic parameters from
+/// the impl with inference variables.
+pub fn types_may_unify(obligation_ty: Ty<'_>, impl_ty: Ty<'_>) -> bool {
+    // Start by checking whether the type in the impl may unify with
+    // pretty much everything. Just return `true`  in that case.
+    match impl_ty.kind() {
+        ty::Param(_) | ty::Projection(_) | ty::Error(_) => return true,
+        ty::Bool
+        | ty::Char
+        | ty::Int(_)
+        | ty::Uint(_)
+        | ty::Float(_)
+        | ty::Adt(..)
+        | ty::Str
+        | ty::Array(..)
+        | ty::Slice(..)
+        | ty::RawPtr(..)
+        | ty::Dynamic(..)
+        | ty::Ref(..)
+        | ty::Never
+        | ty::Tuple(..)
+        | ty::FnPtr(..)
+        | ty::Foreign(..)
+        | ty::Opaque(..) => {}
+        ty::FnDef(..)
+        | ty::Closure(..)
+        | ty::Generator(..)
+        | ty::GeneratorWitness(..)
+        | ty::Placeholder(..)
+        | ty::Bound(..)
+        | ty::Infer(_) => bug!("unexpected impl_ty: {impl_ty}"),
+    }
+
+    let k = impl_ty.kind();
+    match *obligation_ty.kind() {
+        // Purely rigid types, use structural equivalence.
+        ty::Bool
+        | ty::Char
+        | ty::Int(_)
+        | ty::Uint(_)
+        | ty::Float(_)
+        | ty::Str
+        | ty::Never
+        | ty::Foreign(_) => obligation_ty == impl_ty,
+        ty::Ref(_, obl_ty, obl_mutbl) => match k {
+            &ty::Ref(_, impl_ty, impl_mutbl) => {
+                obl_mutbl == impl_mutbl && types_may_unify(obl_ty, impl_ty)
+            }
+            _ => false,
+        },
+        ty::Adt(obl_def, obl_substs) => match k {
+            &ty::Adt(impl_def, impl_substs) => {
+                obl_def == impl_def
+                    && iter::zip(obl_substs, impl_substs)
+                        .all(|(obl, imp)| generic_args_may_unify(obl, imp))
+            }
+            _ => false,
+        },
+        ty::Slice(obl_ty) => matches!(k, &ty::Slice(impl_ty) if types_may_unify(obl_ty, impl_ty)),
```

Comment:
You could use `matches!`-with-an-`if` in more of these cases... probably good to have consistency one way or the other. Since this is the only one at the moment, maybe a `match` would be better here?

Reply:
my issue with matches is that it somewhat breaks formatting which is annoying.

Making this consistent is a bit annoying :laughing: going to keep it this way for now '^^